File: week14/11/5/P5.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

class Solution {
public:
    vector<int> pancakeSort(vector<int>& arr) {
        vector<int> result;
        helper(result, arr);
        return result;
    }
    
    void helper(vector<int>& result, vector<int>& arr) {
        if (arr.size() == 1) return;
        
        // Find the index of the maximum element
        int maxIndex = max_element(arr.begin(), arr.end()) - arr.begin();
        
        // If the maximum element is not in the correct position
        if (maxIndex != arr.size() - 1) {
            if (maxIndex != 0) {
                // Flip the part of the array that contains the maximum element to move it to the front
                result.push_back(maxIndex + 1);
                reverse(arr.begin(), arr.begin() + maxIndex + 1);
            }
            
            // Flip the entire array to move the maximum element to its correct position
            result.push_back(arr.size());
            reverse(arr.begin(), arr.end());
        }
        
        // Recursively call helper with the reduced array size (excluding the last element)
        arr.pop_back();
        helper(result, arr);
    }
};

int main() {
    Solution solution;
    vector<int> arr = {3, 2, 4, 1};
    
    vector<int> result = solution.pancakeSort(arr);
    
    // Print the result
    for (int flip : result) {
        cout << flip << " ";
    }
    cout << endl;

    return 0;
}

File: week14/11/5/P5.py
class Solution:
    def pancakeSort(self, arr: List[int]) -> List[int]:
        return self.helper([], arr)
    
    def helper(self, result, arr):
        if len(arr) == 1: 
            return result
        
        b = arr.index(max(arr)) 
        if b != len(arr) - 1:
            if b != 0: 
                result.append(b + 1)
                arr = arr[:b+1][::-1] + arr[b+1:]  # 첫 번째 플립

            result.append(len(arr))  # 최대값을 올바른 위치로 이동
            arr = arr[::-1]  # 두 번째 플립

        return self.helper(result, arr[:-1])  # 배열 크기 축소 후 재귀 호출